src/webgpu/util/floating_point.ts

```diff
@@ -0,0 +1,3090 @@
+  impl: MatrixToMatrix;
+}
+
+/**
+ * A function that converts a pair of matrices to a matrix of acceptance
+ * intervals.
+ * This is the public facing API for builtin implementations that is called
+ * from tests.
+ */
+export interface MatrixPairToMatrix {
+  (x: Matrix<number>, y: Matrix<number>): FPMatrix;
+}
+
+/**
+ * A function that converts a matrix and a scalar to a matrix of acceptance
+ * intervals.
+ * This is the public facing API for builtin implementations that is called
+ * from tests.
+ */
+export interface MatrixScalarToMatrix {
+  (x: Matrix<number>, y: number): FPMatrix;
+}
+
+/**
+ * A function that converts a scalar and a matrix to a matrix of acceptance
+ * intervals.
+ * This is the public facing API for builtin implementations that is called
+ * from tests.
+ */
+export interface ScalarMatrixToMatrix {
+  (x: number, y: Matrix<number>): FPMatrix;
+}
+
+/**
+ * A function that converts a matrix and a vector to a vector of acceptance
+ * intervals.
+ * This is the public facing API for builtin implementations that is called
+ * from tests.
+ */
+export interface MatrixVectorToVector {
+  (x: Matrix<number>, y: number[]): FPVector;
+}
+
+/**
+ * A function that converts a vector and a matrix to a vector of acceptance
+ * intervals.
+ * This is the public facing API for builtin implementations that is called
+ * from tests.
+ */
+export interface VectorMatrixToVector {
+  (x: number[], y: Matrix<number>): FPVector;
+}
+
+// Context
+
+/**
+ * Typed structure containing all the limits/constants defined for each
+ * WGSL floating point kind
+ */
+interface FPConstants {
+  positive: {
+    min: number;
+    max: number;
+    infinity: number;
+    nearest_max: number;
+    less_than_one: number;
+    subnormal: {
+      min: number;
+      max: number;
+    };
+    pi: {
+      whole: number;
+      three_quarters: number;
+      half: number;
+      third: number;
+      quarter: number;
+      sixth: number;
+    };
+    e: number;
+  };
+  negative: {
+    min: number;
+    max: number;
+    infinity: number;
+    nearest_min: number;
+    less_than_one: number;
+    subnormal: {
+      min: number;
+      max: number;
+    };
+    pi: {
+      whole: number;
+      three_quarters: number;
+      half: number;
+      third: number;
+      quarter: number;
+      sixth: number;
+    };
+  };
+  anyInterval: FPInterval;
+  zeroInterval: FPInterval;
+  negPiToPiInterval: FPInterval;
+  greaterThanZeroInterval: FPInterval;
+  zeroVector: {
+    2: FPVector;
+    3: FPVector;
+    4: FPVector;
+  };
+  anyVector: {
+    2: FPVector;
+    3: FPVector;
+    4: FPVector;
+  };
+  anyMatrix: {
+    2: {
+      2: FPMatrix;
+      3: FPMatrix;
+      4: FPMatrix;
+    };
+    3: {
+      2: FPMatrix;
+      3: FPMatrix;
+      4: FPMatrix;
+    };
+    4: {
+      2: FPMatrix;
+      3: FPMatrix;
+      4: FPMatrix;
+    };
+  };
+}
+
+abstract class FPContext {
```

Comment:
`Context` at least to me is something that carries some form of state. `FPInfo`, `FPTraits` ? Naming things is hard :).

AFAICT, this class is entirely stateless. It carries the `FPKind`, which is immutable. I probably would have gone with an interface here, just listing all the `public` members. I don't know JS well enough to know if there's any real pros or cons to using a class.

Reply:
I agree context implies a stateful thing that is pass around . I think I will rename this to be FPTraits, since it is telling the system about the traits that distinguish each floating point type. (My first choice would have been FPType, but whoa boy that seems like it is going to clash a lot with things mentally)

wrt class v interface.
There is a lot of generic code for calculating intervals that lives in the abstract class here, some which has been factored into multiple functions to encapsulate and hide the details, so uses protected/private a lot. I don't think interfaces really support having private/protected members.
As is, since there is only one FP kind, this class looks like a lot of deck chair organization, but I am hoping when it comes to implementing abstract and f16 the actual implementation is going to be as terse as the f32 section, just defining the constants and specific utilities (other then maybe having to add in a few new constants for things that are magic number currently)

I think there is possibility of refactoring FPTraits to be an interface and having all of the generic framework code live off in a separate FPFramework or something object, but I think that would be a later refinement of what I have here if we feel we need it. I think the focus here is to organize the pile of free functions that I had into a structure that allows expressing new floating point types with minimum code duplication.

Reply:
SGTM